File: _submissions/19UCS159_Tharun/Day 1/Q2.java

/*
Prime Sum : https://www.interviewbit.com/problems/prime-sum/

Given an even number ( greater than 2 ), return two prime numbers whose sum will be equal to given number.
*/
import java.util.*;

public class Q2 {
  public int[] primesum(int A) {
    boolean[] isPrime = new boolean[A + 1];
    Arrays.fill(isPrime, true);
    isPrime[0] = isPrime[1] = false;
    for (int i = 2; i * i <= A; i++) {
      if (isPrime[i]) {
        for (int j = i * i; j <= A; j += i) {
          isPrime[j] = false;
        }
      }
    }
    for (int p = 2; p <= A / 2; p++) {
      if (isPrime[p] && isPrime[A - p]) {
        return new int[] { p, A - p };
      }
    }
    return null;
  }
}

File: _submissions/19UCS125_Rahul/Day1/q2.cpp
vector<int> ans(2, 0);
vector<int> Solution::primesum(int A) {
    if(A % 2 != 0) {
        ans[0] = 2;
        ans[1] = A - 2;
    }
    else {
        vector<bool> prime(A, true);
        prime[0] = prime[1] = false;
        for(int i=2; i*i < A; ++i)
        {
            if(prime[i]==true)
            {
                for(int p=i*i; p<A; p+=i)
                {
                    prime[p]=false;
                }
            }
        }
        for(int i=2; i<A; ++i) {
            if(prime[i] && prime[A-i]) {
                ans[0] = i;
                ans[1] = A - i;
                break;
            }
        }
    }
    return ans;
}